src/ruby/lib/grpc/generic/interceptor_registry.rb

```diff
@@ -0,0 +1,121 @@
+# Copyright 2015 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# GRPC contains the General RPC module.
+module GRPC
+  ##
+  # Represents a registry of added interceptors available for enumeration.
+  # The registry can be used for both server and client interceptors.
+  #
+  class InterceptorRegistry
+    ##
+    # An error raised when an interceptor is attempted to be added
+    # that does not extend GRPC::Interceptor
+    #
+    class DescendantError < StandardError; end
+
+    ##
+    # Initialize the registry with an empty interceptor list
+    #
+    def initialize(interceptors = {})
+      @interceptors = {}
+      interceptors.each do |k, i|
+        self[k] = i
+      end
+    end
+
+    ##
+    # Add an interceptor to the registry
+    #
+    # @param [Symbol] name The key to identify the interceptor in the
+    #   registry as
+    # @param [GRPC::Interceptor] interceptor The interceptor instance
+    #
+    def []=(name, interceptor = nil)
```

Comment:
It looks like the `[]=` method can potentially be bypassed, with bad types accidentally invoked.

AFAICS the features of this class are also offered in `Hash`? 

What do you think if we shrink the `InterceptorRegistry` class here, move the type-checking logic into the `InterceporRegistry#initialize` method, and keep only the `build_context` method. 

The user API would then be to pass in a plain `Hash` to `#new` of this class